pkg/standalone-utils/contracts/StableLPOracle.sol

```diff
@@ -0,0 +1,212 @@
+import { SafeCast } from "@openzeppelin/contracts/utils/math/SafeCast.sol";
+
+import { ILPOracleBase } from "@balancer-labs/v3-interfaces/contracts/standalone-utils/ILPOracleBase.sol";
+import { IStablePool } from "@balancer-labs/v3-interfaces/contracts/pool-stable/IStablePool.sol";
+import { Rounding } from "@balancer-labs/v3-interfaces/contracts/vault/VaultTypes.sol";
+import { IBasePool } from "@balancer-labs/v3-interfaces/contracts/vault/IBasePool.sol";
+import { IVault } from "@balancer-labs/v3-interfaces/contracts/vault/IVault.sol";
+
+import { InputHelpers } from "@balancer-labs/v3-solidity-utils/contracts/helpers/InputHelpers.sol";
+import { FixedPoint } from "@balancer-labs/v3-solidity-utils/contracts/math/FixedPoint.sol";
+import { StableMath } from "@balancer-labs/v3-solidity-utils/contracts/math/StableMath.sol";
+
+import { LPOracleBase } from "./LPOracleBase.sol";
+
+contract StableLPOracle is LPOracleBase {
+    using FixedPoint for uint256;
+    using SafeCast for *;
+
+    // Thrown when the k parameter did not converge to the positive root.
+    error KDidntConverge();
+
+    int256 private constant _ONE_INT = 1e18;
+    int256 private constant _K_MAX_ERROR = 0.000000000001e16; // 0.000000000001%
+
+    constructor(
+        IVault vault_,
+        IStablePool pool_,
+        AggregatorV3Interface[] memory feeds,
+        uint256 version_
+    ) LPOracleBase(vault_, IBasePool(address(pool_)), feeds, version_) {
+        // solhint-disable-previous-line no-empty-blocks
+    }
+
+    /// @inheritdoc ILPOracleBase
+    function calculateTVL(int256[] memory prices) public view override returns (uint256 tvl) {
+        (, , , uint256[] memory lastBalancesLiveScaled18) = _vault.getPoolTokenInfo(address(pool));
+
+        // The TVL of the stable pool is computed by calculating the balances for the stable pool that would represent
+        // the given price vector. To compute these balances, we need only the amplification parameter of the pool,
+        // the invariant and the price vector.
+
+        uint256 invariant = pool.computeInvariant(lastBalancesLiveScaled18, Rounding.ROUND_DOWN);
+
+        uint256[] memory marketPriceBalancesScaled18 = _computeMarketPriceBalances(invariant, prices);
+
+        tvl = 0;
+        for (uint256 i = 0; i < _totalTokens; i++) {
+            tvl += prices[i].toUint256().mulDown(marketPriceBalancesScaled18[i]);
+        }
+
+        return tvl;
+    }
+
+    /**
+     * @dev Computes a set of balances for a given price vector and a given invariant. The set of balances should meet
+     * two conditions:
+     * 1. The invariant of the computed balances should be equal to the given invariant.
+     * 2. The spot prices of the computed balances should be equal to the given price vector. In other words:
+     *    `spotPrice(i, j) == prices[i] / prices[j]`
+     */
+    function _computeMarketPriceBalances(
```

Comment:
This is only called once, during TVL calculation. Consider having this also return the TVL, since it already takes the prices as an argument, and tvl is just the sum of the product of this argument and what this function is already computing. Would save a little gas by eliminating a second loop.

Reply:
I wanted to split them so I can use the balances to compute the current invariant and current spot prices in a test. I mean, the TVL is not predictable, but the spot prices and the invariant are.

Reply:
yeah I think that's fine.